src/field/HelixStepper.hh

```diff
@@ -0,0 +1,164 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2020-2022 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file HelixStepper.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include "base/Types.hh"
+
+#include "Types.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * The Helix Stepper
+ *
+ * The analytical solution for the motion of a charged particle in a uniform
+ * magnetic field.
+ */
+template<class EquationT>
+class HelixStepper
+{
+  public:
+    //!@{
+    //! Type aliases
+    using Result = StepperResult;
+    //!@}
+
+  public:
+    // Construct with the equation of motion
+    CELER_FUNCTION
+    HelixStepper(const EquationT& eq) : equation_(eq) {}
+
+    // Adaptive step size control
+    CELER_FUNCTION auto operator()(real_type step, const OdeState& beg_state)
+        -> Result;
+
+  private:
+    //// DATA ////
+
+    // Equation of the motion
+    const EquationT& equation_;
+
+    //// HELPER FUNCTIONS ////
+
+    // Analytical solution for a given step along a helix trajectory
+    CELER_FUNCTION OdeState move(real_type       step,
+                                 real_type       radius,
+                                 real_type       helicity,
+                                 const OdeState& beg_state,
+                                 const OdeState& rhs);
+
+    //// COMMON PROPERTIES ////
+
+    static CELER_CONSTEXPR_FUNCTION real_type tolerance() { return 1e-10; }
+};
+
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+/*!
+ * An explicit helix stepper with analytical solutions at the end and the
+ * middle point for a given step.
+ *
+ * Assuming that the magnetic field is uniform and chosen to be parallel to
+ * the z-axis, \f$B = (0, 0, B_z)\f$, without loss of generality, the motion of
+ * a charged particle is described by a helix trajectory. For this algorithm,
+ * the radius of the helix, \f$R = m gamma v/(qB)\f$ and the helicity, defined
+ * as \f$ -sign(q B_z)\f$ are evaluated through the right hand side of the ODE
+ * equation where q is the charge of the particle.
+ */
+template<class E>
+CELER_FUNCTION auto
+HelixStepper<E>::operator()(real_type step, const OdeState& beg_state)
+    -> Result
+{
+    Result result;
+
+    // Evaluate the right hand side of the equation
+    OdeState rhs = equation_(beg_state);
+
+    // Calculate the radius of the helix
+    real_type radius
+        = sqrt(ipow<2>(norm(beg_state.mom)) - ipow<2>(beg_state.mom[2]))
```

Comment:
Use `std::` namespace, and avoid the needless `sqrt` inside of `norm`  by using `dot_product`.
```suggestion
        = std::sqrt(dot_product(beg_state.mom, beg_state.mom) - ipow<2>(beg_state.mom[2]))
```

Also, unless you rename this `ZHelixStepper` or something, I think it's a mistake to hard-code Z axes everywhere here. Unless you want to support time-dependent uniform magnetic fields, you should de-template this class and explicitly use `UniformMagField` as the equation type and add assertions that the x/y components are zero, or (better) define a `UniformZMagField` class that just has a scalar value for the magnitude+sign.

Reply:
Good catch!  How come I did this (sqrt, then square)?  Of course, not thinking enough!  Thanks.

For the comment, I see your point, but I cannot imagine that anyone (HEP experiment) ever uses a constant (i.e., homogeneous and static) field in a way different from a single value set along an axis - normally the Helix trajectory makes a good sense in the cylindrical coordinate system where the magnetic field is naturally aligned to the z-direction.  In addition, we can always transform the direction (momentum vector) of a particle with respect to the field direction which is aligned to an axis.   Another alternative may be just to add an accessor to get the field from MagFieldEquation, which will not change any structural change. Anyway, I do not think that Helix stepper will be ever used for a general purpose, other than  other test purposes or predicting an approximated trajectory in a very slowly varying field.  So, if you do not mind, I would like to keep the hard-assumption (z-field) for now (which simplify calculations without changing the current design).

Reply:
I'm 100% OK with with keeping this class simple and not trying to generalize it. However, I think the code design should reflect that it only applies to a very specific field equation (uniform field along Z axis). Adding a simple helper `MagField` should be almost zero work. It looks like the `MagFieldTraits` currently assume that the `Stepper` is templated on a field type, so rather than refactor that we could leave `HelixStepper` as is and add
```c++
static_assert(std::is_same<EquationT, UniformZMagField>::value, "Helix stepper only works with uniform magnetic fields along `z`");
```
to the class definition. This would codify the *assumption* that HelixStepper is only used for uniform-along-Z magnetic fields, which is what I'm looking for. (Un-codified assumptions is one major source of bugs and frustration.)

Reply:
Stepper is templated on a field equation, not on  a field type - the field equation is templated on a field type.  So,  can you elaborate how to propagate a field type to the stepper and check the type? Should I do something similar inside MafFieldTraits instead?

Reply:
Oh, I see. We probably need to add a type alias to `MagFieldEquation`, and then you could make the assertion on `EquationT::Field`?

Reply:
I renamed `HelixStepper` to `ZHelixStepper ` as a short turn-around since there is an additional inconsistency to use  UniformZMagField with a scalar field inside MagFieldEquation which evaluates the right hand side of the equation with a field vector (i.e, cross product of momentum x field_vector)  - using rhs.mom inside the stepper will not be meaningful as MagFieldEquation cannot handle the scalar field as it is currently implemented.  Adding another template argument is also needed for SteppersTest::run_stepper if we introduce UniformZMagField (which is easy to extend, but needs extra redundancy to construct a different field inside each test case).

Reply:
I added `UniformZMagField `and a check with the field type for constructing `ZHelixStepper`, and modified tests accordingly, which should address all your suggestions.